src/ipc/Queue.h

```diff
@@ -410,6 +426,96 @@ OneToOneUniQueue::push(const Value &value, QueueReader *const reader)
     return wasEmpty && (!reader || reader->raiseSignal());
 }
 
+template <class Value>
+void
+OneToOneUniQueue::statIn(StoreEntry &entry) const
+{
+    // Nobody can modify our theIn so, after capturing some valid theSize value
+    // in count, we can reliably report all [theIn-count, theIn) items that were
+    // queued at theSize capturing time. We may report items already pop()ed by
+    // the other side, but that is OK because pop() does not modify items -- it
+    // only increments theOut.
```

Comment:
This comment talks about 'poping by the other side' but 'theIn' index is about 'input', which changes by the other side push(). Don't we need to move this comment into statOut()?

Reply:
Good catch! I agree that there are problems here. However, the (lying) comment is actually placed correctly. It would make no sense to justify theIn usage in a method that uses theOut (and vice versa). 

FWIW, my confusion stemmed from the clash between "in" in theIn and "in" in statIn()/inQueue(). The former is a push() index/counter -- it counts things going into the queue. The latter indicates an incoming/receiving queue (from the current process point of view). I was trying to justify the broken code (which I assumed was not broken), resulting in a conflict you thankfully found.

My statIn() comment claims that statIn() caller owns/controls theIn. That claim allows the statIn() method to use theIn safely (as the comment explains). That claim itself is actually a lie! The caller (i.e. the receiver of queued messages) owns/controls theOut instead. The statIn() code is broken -- it is using a non-atomic shared memory variable (theIn) that is being modified by another process. Same for statOut(), of course.

AFAICT, statIn() code (including its comment) should be swapped with the statOut() code (including its comment). Similarly, `outputs writer statistics` should be changed to `reports incoming queue state` and `outputs reader statistics` to `reports outgoing queue state`.

I was actually going to move printing of the "kid receiving from..." and "kid sending from..." headers into the statIn/Out() methods to clarify this in code, but I ran out of time. It requires adding/passing a remoteProcessId parameter to statIn/Out(), but I think it may be worth it.

I probably would not rename statIn() to statIncoming() to keep the symmetry with inQueue() and such. All these names should probably be spelled out or changed to address the root of the problem, but that would cause too many out-of-scope changes. I would not do it now/here.

Reply:
Done at 47365c1 and 5db9af2.